File: apps/gproxy/frontend/src/modules/admin/providers/channels/vertex/credential.ts
import type { ChannelCredentialSchema } from "../../types";

export const CREDENTIAL_SCHEMA: ChannelCredentialSchema = {
  "channel": "vertex",
  "kind": "builtin/vertex",
  "wrapper": "Builtin",
  "builtinVariant": "Vertex",
  "fields": [
    {
      "key": "project_id",
      "label": "project_id",
      "type": "string"
    },
    {
      "key": "client_email",
      "label": "client_email",
      "type": "string"
    },
    {
      "key": "private_key",
      "label": "private_key",
      "type": "string"
    },
    {
      "key": "private_key_id",
      "label": "private_key_id",
      "type": "string"
    },
    {
      "key": "client_id",
      "label": "client_id",
      "type": "string"
    },
    {
      "key": "auth_uri",
      "label": "auth_uri",
      "type": "optional_string"
    },
    {
      "key": "token_uri",
      "label": "token_uri",
      "type": "optional_string"
    },
    {
      "key": "auth_provider_x509_cert_url",
      "label": "auth_provider_x509_cert_url",
      "type": "optional_string"
    },
    {
      "key": "client_x509_cert_url",
      "label": "client_x509_cert_url",
      "type": "optional_string"
    },
    {
      "key": "universe_domain",
      "label": "universe_domain",
      "type": "optional_string"
    },
    {
      "key": "access_token",
      "label": "access_token",
      "type": "string"
    },
    {
      "key": "expires_at",
      "label": "expires_at",
      "type": "integer"
    }
  ]
};

File: crates/gproxy-provider/src/channels/vertex/credential.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct VertexServiceAccountCredential {
    pub project_id: String,
    pub client_email: String,
    pub private_key: String,
    pub private_key_id: String,
    pub client_id: String,
    pub auth_uri: Option<String>,
    pub token_uri: Option<String>,
    pub auth_provider_x509_cert_url: Option<String>,
    pub client_x509_cert_url: Option<String>,
    pub universe_domain: Option<String>,
    #[serde(default)]
    pub access_token: String,
    #[serde(default)]
    pub expires_at: i64,
}

impl Default for VertexServiceAccountCredential {
    fn default() -> Self {
        Self {
            project_id: String::new(),
            client_email: String::new(),
            private_key: String::new(),
            private_key_id: String::new(),
            client_id: String::new(),
            auth_uri: Some("https://accounts.google.com/o/oauth2/auth".to_string()),
            token_uri: Some("https://oauth2.googleapis.com/token".to_string()),
            auth_provider_x509_cert_url: None,
            client_x509_cert_url: None,
            universe_domain: None,
            access_token: String::new(),
            expires_at: 0,
        }
    }
}

impl VertexServiceAccountCredential {
    pub fn has_refresh_material(&self) -> bool {
        !self.project_id.trim().is_empty()
            && !self.client_email.trim().is_empty()
            && !self.private_key.trim().is_empty()
    }
}
